File: pixleesdk/src/main/java/com/pixlee/pixleesdk/ui/adapter/ProductAdapter.java
package com.pixlee.pixleesdk.ui.adapter;

import android.view.View;
import android.view.ViewGroup;

import androidx.recyclerview.widget.RecyclerView;

import com.pixlee.pixleesdk.PXLProduct;
import com.pixlee.pixleesdk.ui.viewholder.ProductViewHolder;

import java.util.HashMap;
import java.util.List;

/**
 * Created by sungjun on 6/2/20.
 */
public class ProductAdapter extends RecyclerView.Adapter<ProductViewHolder> {
    /**
     * String: product id
     * Boolean: is bookmarked
     */
    HashMap<String, Boolean> bookmarkMap;
    public List<PXLProduct> list;

    public interface ProductListener {
        void onClicked(PXLProduct product);
    }

    ProductListener listener;

    public ProductAdapter(List<PXLProduct> list, HashMap<String, Boolean> bookmarkMap, ProductListener listener) {
        this.list = list;
        this.bookmarkMap = bookmarkMap;
        this.listener = listener;
    }

    public ProductViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return ProductViewHolder.create(parent);
    }

    @Override
    public void onBindViewHolder(final ProductViewHolder holder, int position) {
        PXLProduct product = list.get(position);
        Boolean isBookmarked = null;
        if (bookmarkMap != null) {
            isBookmarked = bookmarkMap.get(product.id);
        }

        holder.bind(product, isBookmarked);
        holder.binding.cardView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                listener.onClicked(list.get(holder.getAdapterPosition()));
            }
        });
    }

    @Override
    public int getItemCount() {
        return list.size();
    }
}

File: pixleesdk/src/main/java/com/pixlee/pixleesdk/ui/adapter/ProductAdapter.kt
package com.pixlee.pixleesdk.ui.adapter

import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.pixlee.pixleesdk.data.PXLProduct
import com.pixlee.pixleesdk.ui.viewholder.ProductViewHolder
import com.pixlee.pixleesdk.ui.viewholder.ProductViewHolder.Companion.create
import java.util.*

/**
 * This is to be used in PXLPhotoRecyclerView which extends RecyclerView
 */
class ProductAdapter(
        val configuration: ProductViewHolder.Configuration,
        val list: List<PXLProduct>,
                     /**
                      * String: product id
                      * Boolean: is bookmarked
                      */
                     val bookmarkMap: HashMap<String, Boolean>?,
                     val onBookmarkChanged: (productId: String, isBookmarkChecked: Boolean) -> Unit,
                     val onItemClicked: (product: PXLProduct) -> Unit) : RecyclerView.Adapter<ProductViewHolder?>() {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ProductViewHolder {
        return create(parent)
    }

    override fun onBindViewHolder(holder: ProductViewHolder, position: Int) {
        val product = list[position]
        if (bookmarkMap!=null) {
            holder.bind(product, bookmarkMap[product.id] ?: false, configuration)
        }else{
            holder.bind(product, null, configuration)
        }
        holder.binding.bookmark.setOnClickListener {
            val productId = list[holder.adapterPosition].id
            bookmarkMap?.also {
                val bookmarked = it[productId] ?: false
                it[productId] = !bookmarked
                onBookmarkChanged(list[holder.adapterPosition].id, !bookmarked)
                holder.changeBookmarkUI(!bookmarked, configuration.bookmarkDrawable)
            }
        }

        holder.itemView.setOnClickListener {
            onItemClicked(list[holder.adapterPosition])
        }
    }

    override fun getItemCount(): Int {
        return list.size
    }
}